### DISCUSSION TITLE
Struggling with DI in a worker project

### DISCUSSION THREAD

**Discussion by lwmorris067** [MarimerLLC/csla#3071] [Questions] [answered]
I've been able to get a new Blazor application setup and using Csla 6.x.

I've struggled the past several days getting a worker process to work. I don't know what I'm doing, so it's probably something simple. Any help would be greatly appreciated. When I run the worker, I get the following error in program.cs. It occures for the IHost host = block of code.

"System.AggregateException: 'Some services are not able to be constructed (Error while validating the service descriptor 'ServiceType: Microsoft.Extensions.Hosting.IHostedService Lifetime: Singleton ImplementationType: BlazorApp.Worker.Worker': Cannot consume scoped service 'Csla.ApplicationContext' from singleton 'Microsoft.Extensions.Hosting.IHostedService'.)'"

Thank you any help!

Program.cs looks like
```
using Csla.Configuration;
using BlazorApp.Worker;

IHost host = Host.CreateDefaultBuilder(args)
    .ConfigureServices(services =>
    {
        services.AddHostedService<Worker>();
        services.AddOptions();

        services.AddCsla(o => o
            .DataPortal(dpo => dpo.UseLocalProxy())
        );
    })
    .Build();

await host.RunAsync();
```

Worker.cs looks like

```
namespace BlazorApp.Worker
{
    public class Worker : BackgroundService
    {
        public Worker(ILogger<Worker> logger, Csla.IDataPortal<BlazorApp1.Shared.PersonEdit> personPortal)
        {
        }

        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
        {
            while (!stoppingToken.IsCancellationRequested)
            {
            }
        }
    }
}
```

**Comment by lwmorris067** [comment, marked as answer]
Changing Worker.cs to this works. Seems cheesy, though. DI passes in IServiceProvider, I CreateScope() and grab the ServiceProvider off of it. THEN I can use the dataportal.

meh.

```
using Csla;
using Csla.Configuration;
using Csla.Runtime;

namespace BlazorApp.Worker
{
    public class Worker : BackgroundService
    {
        private IServiceProvider _provider;

        public Worker(ILogger<Worker> logger, IServiceProvider p)
        {
            _provider = p.CreateScope().ServiceProvider;

            var works = _provider.GetService<IDataPortal<BlazorApp1.Shared.PersonEdit>>();
            var bo = works.Create();
        }

        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
        {
            while (!stoppingToken.IsCancellationRequested)
            {
            }
        }
    }
}
```

**Reply by lwmorris067** [reply]
After a little more reading, it sounds like the CreateScope method is maybe how it's actually supposed to work.

**Reply by rockfordlhotka** [reply]
Yes, when running in aspnetcore, CSLA assumes it is running within a DI scope (as that's the normal scenario). I think it is interesting that a worker doesn't run in a DI scope like everything else in aspnetcore.
